xdsl/transforms/csl_stencil_bufferize.py

```diff
@@ -371,6 +290,104 @@ def match_and_rewrite(self, op: arith.Constant, rewriter: PatternRewriter, /):
         )
 
 
+@dataclass(frozen=True)
+class LinalgAccumulatorInjection(RewritePattern):
+    """
+    Injects the `accumulator` block argument for linalg ops within the csl_stencil.apply regions
+    into the `outs` argument. This typically reduces the overhead of bufferization.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: linalg.NamedOpBase | linalg.Generic, rewriter: PatternRewriter, /
+    ):
+        # check if there is an output to inject the accumulator into
+        if len(op.outputs) != 1 or not isa(
+            target_t := op.outputs[0].type, AnyTensorType
+        ):
+            return
+
+        # find parent `csl_stencil.apply` and which of the regions `op` is in
+        p_region = op.parent_region()
+        apply = None
+        while (
+            p_region
+            and (apply := p_region.parent_op())
+            and not isinstance(apply, csl_stencil.ApplyOp)
+        ):
+            p_region = apply.parent_region()
+        if not isinstance(apply, csl_stencil.ApplyOp) or not p_region:
+            return
+
+        # retrieve the correct accumulator block arg
+        if p_region == apply.receive_chunk:
+            acc_block_arg = p_region.block.args[2]
+        elif p_region == apply.done_exchange:
+            acc_block_arg = p_region.block.args[1]
+        else:
+            raise ValueError("Invalid region")
+
+        # fetch the bufferization of the accumulator block arg
+        acc_uses = [
+            use.operation
+            for use in acc_block_arg.uses
+            if isinstance(use.operation, bufferization.ToTensorOp)
+        ]
+        if len(acc_uses) < 1:
+            return
+        acc_bufferization = acc_uses[0]
+
+        # in the `chunk_recieved` region, fetch or create a down-sized chunk of the accumulator
+        if (
+            acc_bufferization.tensor.type != target_t
+            and p_region == apply.receive_chunk
+        ):
+            # check if we can find an `extract_slice` to the desired size
+            extract_slice = None
+            for use in acc_bufferization.tensor.uses:
+                if (
+                    isinstance(use.operation, tensor.ExtractSliceOp)
+                    and use.operation.result.type == target_t
+                ):
+                    extract_slice = use.operation
+                    break
+
+            # create `extract_slice` op if none exists
+            if not extract_slice:
```

Comment:
I don't really get this part. Why do we make an `extract_slice` rather than not doing anything?

Reply:
This part actually existed, but the logic got refactored. When communicating in chunks, the `extract_slice` op is needed for bufferization to be able to use the accumulator as an accumulator. Even though there is an insert_slice op later, which logically is all we need, bufferization does not draw the implication that it could be using the slice as an accumulator before inserting there.